File: frontend/lib/features/task/domain/entities/task.dart
import '../../../tag/domain/entities/tag.dart';

/// Mirrors the backend enums; JSON wire names live in the data layer.
enum TaskStatus { todo, inProgress, done }

enum TaskPriority { low, medium, high }

final class Task {
  const Task({
    required this.id,
    required this.title,
    required this.status,
    required this.priority,
    required this.createdAt,
    required this.updatedAt,
    this.description,
    this.dueDate,
    this.listId,
    this.tags = const [],
  });

  final String id;
  final String title;
  final String? description;
  final TaskStatus status;
  final TaskPriority priority;
  final DateTime? dueDate;
  final String? listId;
  final List<Tag> tags;
  final DateTime createdAt;
  final DateTime updatedAt;

  Task copyWith({
    String? title,
    String? description,
    TaskStatus? status,
    TaskPriority? priority,
    DateTime? dueDate,
    String? listId,
    bool clearListId = false,
    List<Tag>? tags,
  }) {
    return Task(
      id: id,
      title: title ?? this.title,
      description: description ?? this.description,
      status: status ?? this.status,
      priority: priority ?? this.priority,
      dueDate: dueDate ?? this.dueDate,
      listId: clearListId ? null : (listId ?? this.listId),
      tags: tags ?? this.tags,
      createdAt: createdAt,
      updatedAt: updatedAt,
    );
  }
}

File: backend/src/main/java/com/smarttodo/domain/model/Task.java
package com.smarttodo.domain.model;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

/**
 * A task owned by a user. Pure domain object: no framework imports.
 * May belong to at most one list and carry any number of tags.
 * completedAt tracks when the task entered DONE (for analytics).
 */
public record Task(
		UUID id,
		UUID userId,
		String title,
		String description,
		TaskStatus status,
		TaskPriority priority,
		Instant dueDate,
		Instant completedAt,
		UUID listId,
		List<Tag> tags,
		Instant createdAt,
		Instant updatedAt) {

	public Task {
		tags = tags == null ? List.of() : List.copyOf(tags);
	}

	public static Task createNew(UUID userId, String title, String description,
			TaskPriority priority, Instant dueDate) {
		Instant now = Instant.now();
		return new Task(UUID.randomUUID(), userId, title, description,
				TaskStatus.TODO, priority == null ? TaskPriority.MEDIUM : priority,
				dueDate, null, null, List.of(), now, now);
	}

	public Task withListAndTags(UUID listId, List<Tag> tags) {
		return new Task(id, userId, title, description, status, priority, dueDate,
				completedAt, listId, tags, createdAt, updatedAt);
	}

	public Task update(String title, String description, TaskStatus status,
			TaskPriority priority, Instant dueDate) {
		return update(title, description, status, priority, dueDate, listId, tags);
	}

	public Task update(String title, String description, TaskStatus status,
			TaskPriority priority, Instant dueDate, UUID listId, List<Tag> tags) {
		return new Task(id, userId, title, description, status, priority, dueDate,
				resolveCompletedAt(status), listId, tags, createdAt, Instant.now());
	}

	/**
	 * completedAt is stamped when entering DONE and cleared when leaving it;
	 * an already-DONE task keeps its original completion time.
	 */
	private Instant resolveCompletedAt(TaskStatus newStatus) {
		if (newStatus != TaskStatus.DONE) {
			return null;
		}
		return completedAt != null ? completedAt : Instant.now();
	}
}
